Xamarin.Essentials/Types/FileProvider.android.cs

```diff
@@ -0,0 +1,102 @@
+            // we want the internal storage
+            if (TemporaryLocation == FileProviderLocation.Internal)
+                return Platform.AppContext.CacheDir;
+
+            // if we need permission, then check
+            // if we can fall back, then don't throw
+            var externalOnly = TemporaryLocation == FileProviderLocation.External;
+            var hasPermission = Platform.HasApiLevel(BuildVersionCodes.Kitkat) || Permissions.EnsureDeclared(PermissionType.ExternalStorage, externalOnly);
+
+            // make sure the external storage is available
+            var hasExternalMedia = Platform.HasApiLevel(BuildVersionCodes.Lollipop)
+                ? AndroidEnvironment.GetExternalStorageState(Platform.AppContext.ExternalCacheDir) == AndroidEnvironment.MediaMounted
+#pragma warning disable CS0618 // Type or member is obsolete
+                : AndroidEnvironment.GetStorageState(Platform.AppContext.ExternalCacheDir) == AndroidEnvironment.MediaMounted;
+#pragma warning restore CS0618 // Type or member is obsolete
+
+            // undo all the work if we have requested a fail (mainly for testing)
+            if (AlwaysFailExternalMediaAccess)
+                hasExternalMedia = false;
+
+            // fail if we need the external storage, but there is none
+            if (externalOnly && !hasExternalMedia)
+                throw new InvalidOperationException("Unable to access the external storage, the media is not mounted.");
+
+            // based on permssions, return the correct directory
+            // if permission were required, then it would have already thrown
+            return hasPermission && hasExternalMedia
+                ? Platform.AppContext.ExternalCacheDir
+                : Platform.AppContext.CacheDir;
+        }
+
+        internal static bool IsFileInPublicLocation(string filename)
+        {
+            // get the Android path, we use "CanonicalPath" instead of "AbsolutePath"
+            // because we want to resolve any ".." and links/redirects
+            var file = new Java.IO.File(filename);
+            filename = file.CanonicalPath;
+
+            // the shared paths from the "xamarin_essentials_fileprovider_file_paths.xml" resource
+            var publicLocations = new[]
+            {
+                AndroidEnvironment.ExternalStorageDirectory.CanonicalPath,
+                Platform.AppContext.ExternalCacheDir.CanonicalPath,
+                Platform.AppContext.CacheDir.CanonicalPath,
+            };
+
+            foreach (var location in publicLocations)
+            {
+                // make sure we have a trailing slash
+                var suffixedPath = filename.EndsWith(Java.IO.File.Separator)
+                    ? filename
+                    : filename + Java.IO.File.Separator;
+
+                // check if the requested file is in a folder
+                if (suffixedPath.StartsWith(location, StringComparison.OrdinalIgnoreCase))
+                    return true;
```

Comment:
There probably is a nice, special way to si if path A is a parent of path B, but here we just do a simple diff.